actionpack/lib/action_controller/metal.rb

```diff
@@ -126,7 +126,7 @@ class Metal < AbstractController::Base
     # ==== Returns
     # * <tt>string</tt>
     def self.controller_name
-      @controller_name ||= name.demodulize.delete_suffix("Controller").underscore
+      @controller_name ||= (name.demodulize.delete_suffix("Controller").underscore unless anonymous?)
```

Comment:
If we care at all about preserving the controller name we could potentially leave this as is and add a method to `ActionView::TestCase::TestController` instead:

```rb
      def self.controller_name
        "test"
      end
```

Reply:
It doesn't seem to matter (at least in our tests) that the name is unchanged. Anonymous controllers _should_ work, and we use this pattern elsewhere https://github.com/rails/rails/blob/854587c268a5474f824215c84409d44466fba9dd/actionpack/lib/abstract_controller/base.rb#L136